File: android/app/src/main/java/com/naif/realityengine/SecurityScanner.java
package com.naif.realityengine;

import java.util.*;
import java.util.regex.*;

public class SecurityScanner {

    // ─── OWASP Map ───────────────────────────────────────
    public static final java.util.Map<String, String> OWASP_MAP = new java.util.HashMap<String, String>() {{
        put("SQL",     "OWASP A03:2021 | CWE-89  | CVSS 9.8");
        put("XSS",     "OWASP A03:2021 | CWE-79  | CVSS 7.2");
        put("SECRET",  "OWASP A02:2021 | CWE-798 | CVSS 9.1");
        put("CMD",     "OWASP A03:2021 | CWE-78  | CVSS 9.8");
        put("CRYPTO",  "OWASP A02:2021 | CWE-327 | CVSS 7.4");
        put("AUTH",    "OWASP A07:2021 | CWE-287 | CVSS 8.1");
        put("LOG",     "OWASP A09:2021 | CWE-532 | CVSS 5.3");
        put("HTTP",    "OWASP A02:2021 | CWE-319 | CVSS 7.5");
    }};

    public static class SecurityIssue {
        public String title;
        public String description;
        public String fix;
        public int line;
        public String severity;

        public SecurityIssue(String title, String desc, String fix, int line, String sev) {
            this.title = title;
            this.description = desc;
            this.fix = fix;
            this.line = line;
            this.severity = sev;
        }
    }

    public static List<SecurityIssue> scan(String code, String fileName) {
        List<SecurityIssue> issues = new ArrayList<>();
        if (code == null || code.isEmpty()) { return issues; }
        String[] lines = code.split("\n");
        String ext = fileName.contains(".") ? fileName.substring(fileName.lastIndexOf('.') + 1).toLowerCase() : "";
        boolean isJS = ext.equals("js") || ext.equals("jsx") || ext.equals("ts");

        for (int i = 0; i < lines.length; i++) {
            String t = lines[i].trim();
            if (t.isEmpty() || t.startsWith("//") || t.startsWith("#") || t.startsWith("*")) { continue; }
            int ln = i + 1;

            // 1. Hardcoded Password
            if (Pattern.compile("(?i)(password|secret|api_key|apikey|token)\\s*=\\s*[\"'][^\"']{4,}[\"']").matcher(t).find()) {
                if (!Pattern.compile("(?i)(test|example|placeholder)").matcher(t).find()) {
                    issues.add(new SecurityIssue("كلمة مرور مُضمَّنة في الكود",
                        "استخدم Environment Variables", "// استخدم BuildConfig أو Secrets", ln, "CRITICAL"));
                }
            }

            // 2. SQL Injection
            if (Pattern.compile("(?i)(query|execute|sql)").matcher(t).find()) {
                if (Pattern.compile("\"\\s*\\+\\s*\\w+").matcher(t).find()) {
                    issues.add(new SecurityIssue("SQL Injection محتمل",
                        "استخدم Prepared Statements", "// استخدم ? placeholders", ln, "CRITICAL"));
                }
            }

            // 3. eval()
            if (Pattern.compile("\\beval\\s*\\(").matcher(t).find()) {
                issues.add(new SecurityIssue("eval() خطير",
                    "eval() يسمح بتنفيذ أي كود", "// احذف eval()", ln, "CRITICAL"));
            }

            // 4. XSS
            if (isJS && Pattern.compile("innerHTML|document\\.write").matcher(t).find()) {
                issues.add(new SecurityIssue("XSS محتمل: innerHTML",
                    "استخدم textContent", t.replace("innerHTML", "textContent"), ln, "HIGH"));
            }

            // 5. HTTP
            if (Pattern.compile("http://(?!localhost|127)").matcher(t).find()) {
                issues.add(new SecurityIssue("HTTP غير مشفر",
                    "استخدم HTTPS", t.replace("http://", "https://"), ln, "MEDIUM"));
            }

            // 6. Log Secrets
            if (Pattern.compile("(?i)(Log\\.|System\\.out|println)").matcher(t).find()) {
                if (Pattern.compile("(?i)(password|token|secret|key)").matcher(t).find()) {
                    issues.add(new SecurityIssue("تسجيل بيانات حساسة",
                        "لا تسجّل secrets في الـ logs", "// احذف هذا السطر", ln, "HIGH"));
                }
            }

            // 7. Command Injection
            if (Pattern.compile("(?i)(Runtime\\.exec|ProcessBuilder)").matcher(t).find()) {
                if (Pattern.compile("\\+").matcher(t).find()) {
                    issues.add(new SecurityIssue("Command Injection محتمل",
                        "لا تدمج input في shell commands",
                        "// استخدم ProcessBuilder مع List<String>", ln, "CRITICAL"));
                }
            }
        }
        return issues;
    }
}

File: vercel-api/public/security_scanner.js
// ─── Security Scanner ────────────────────────────────
// يكتشف ثغرات أمنية شائعة في الكود

function scanSecurity(code, fileName) {
  const issues = [];
  const lines = code.split('\n');
  const ext = fileName.split('.').pop().toLowerCase();

  lines.forEach((line, i) => {
    const t = line.trim();
    if (!t || t.startsWith('//') || t.startsWith('#') || t.startsWith('*')) return;
    const ln = i + 1;

    // ─── 1. SQL Injection ───────────────────────────
    if (/query|execute|sql/i.test(t) && /\+\s*\w+|\$\{/.test(t) && !/prepared|parameterized/i.test(t)) {
      issues.push({
        type: 'security', sev: 'c',
        title: '🔴 SQL Injection محتمل',
        line: ln, ev: t,
        fix: '// استخدم Prepared Statements بدل string concatenation',
        conf: 85, cIcon: '🔴', cAct: 'ثغرة حرجة',
        cEv: ['لا تدمج input المستخدم مباشرة في SQL']
      });
    }

    // ─── 2. Hardcoded Password/Secret ──────────────
    if (/password|secret|api_key|apikey|token/i.test(t) &&
        /=\s*["'][^"']{4,}["']/.test(t) &&
        !/test|example|placeholder|your_/i.test(t)) {
      issues.push({
        type: 'security', sev: 'c',
        title: '🔴 كلمة مرور/مفتاح مُضمَّن في الكود',
        line: ln, ev: t,
        fix: '// استخدم Environment Variables بدل hardcoded secrets',
        conf: 90, cIcon: '🔴', cAct: 'ثغرة حرجة',
        cEv: ['لا تضع secrets في الكود — استخدم .env']
      });
    }

    // ─── 3. XSS ────────────────────────────────────
    if (['js','jsx','ts','html'].includes(ext)) {
      if (/innerHTML|document\.write|\.html\(/.test(t) && /\w+/.test(t)) {
        issues.push({
          type: 'security', sev: 'h',
          title: '🟠 XSS محتمل: innerHTML بدون sanitize',
          line: ln, ev: t,
          fix: t.replace('innerHTML', 'textContent'),
          conf: 80, cIcon: '🟠', cAct: 'ثغرة XSS',
          cEv: ['استخدم textContent أو sanitize الـ input']
        });
      }
    }

    // ─── 4. eval() ─────────────────────────────────
    if (/\beval\s*\(/.test(t)) {
      issues.push({
        type: 'security', sev: 'c',
        title: '🔴 eval() خطير جداً',
        line: ln, ev: t,
        fix: '// احذف eval() — استخدم JSON.parse أو طريقة آمنة',
        conf: 95, cIcon: '🔴', cAct: 'ثغرة حرجة',
        cEv: ['eval() يسمح بتنفيذ أي كود — خطر جداً']
      });
    }

    // ─── 5. Command Injection ───────────────────────
    if (/exec|spawn|system|subprocess/i.test(t) && /\+|f"|f'|\$\{/.test(t)) {
      issues.push({
        type: 'security', sev: 'c',
        title: '🔴 Command Injection محتمل',
        line: ln, ev: t,
        fix: '// لا تدمج input المستخدم في shell commands',
        conf: 85, cIcon: '🔴', cAct: 'ثغرة حرجة',
        cEv: ['استخدم قائمة arguments بدل string']
      });
    }

    // ─── 6. Insecure Random ─────────────────────────
    if (/Math\.random\(\)|random\(\)/.test(t) && /token|password|secret|id|key/i.test(t)) {
      issues.push({
        type: 'security', sev: 'h',
        title: '🟠 Random غير آمن للـ security',
        line: ln, ev: t,
        fix: '// استخدم crypto.randomBytes() أو SecureRandom()',
        conf: 80, cIcon: '🟠', cAct: 'ثغرة أمنية',
        cEv: ['Math.random() ليس آمناً للـ tokens']
      });
    }

    // ─── 7. MD5 / SHA1 ─────────────────────────────
    if (/md5|sha1|sha-1/i.test(t) && /hash|encrypt|password/i.test(t)) {
      issues.push({
        type: 'security', sev: 'h',
        title: '🟠 خوارزمية تشفير ضعيفة: MD5/SHA1',
        line: ln, ev: t,
        fix: '// استخدم SHA-256 أو bcrypt للكلمات السر',
        conf: 85, cIcon: '🟠', cAct: 'تشفير ضعيف',
        cEv: ['MD5/SHA1 ضعيفة — استخدم SHA-256 أو bcrypt']
      });
    }

    // ─── 8. HTTP بدل HTTPS ──────────────────────────
    if (/http:\/\/(?!localhost|127\.0\.0\.1)/.test(t) && !/\/\//i.test(t.slice(0, t.indexOf('http')))) {
      issues.push({
        type: 'security', sev: 'm',
        title: '🟡 HTTP غير مشفر — استخدم HTTPS',
        line: ln, ev: t,
        fix: t.replace('http://', 'https://'),
        conf: 75, cIcon: '🟡', cAct: 'اتصال غير آمن',
        cEv: ['HTTP يرسل البيانات بدون تشفير']
      });
    }

    // ─── 9. Log Sensitive Data ──────────────────────
    if (/console\.log|print|Log\./i.test(t) && /password|token|secret|key|credit/i.test(t)) {
      issues.push({
        type: 'security', sev: 'h',
        title: '🟠 تسجيل بيانات حساسة في الـ logs',
        line: ln, ev: t,
        fix: '// لا تسجّل passwords أو tokens في الـ logs',
        conf: 85, cIcon: '🟠', cAct: 'تسرب بيانات',
        cEv: ['البيانات الحساسة في الـ logs خطر أمني']
      });
    }

    // ─── 10. Weak TLS ───────────────────────────────
    if (/SSLv3|TLSv1\.0|TLSv1\.1|ssl_version|PROTOCOL_TLS/i.test(t)) {
      issues.push({
        type: 'security', sev: 'h',
        title: '🟠 إصدار TLS قديم وغير آمن',
        line: ln, ev: t,
        fix: '// استخدم TLS 1.2 أو 1.3',
        conf: 90, cIcon: '🟠', cAct: 'بروتوكول ضعيف',
        cEv: ['TLS 1.0/1.1 معطوبان — استخدم TLS 1.2+']
      });
    }
  });

  return issues;
}
